[New feature][Performance] Introduce a caching mechanism to ListView to reduce unnecessary layout during scrolling

xu-baolin opened the pull request:
## Document description：
Introduced new property `enableItemExtentsCaching` for `ListView`, the documentation here describes how it works.
```dart
  /// {@template flutter.widgets.list_view.enableItemExtentsCaching}
  /// Whether to cache the main axis extents of all laid out children. The
  /// cached data will be used to calculate which children will be loaded during
  /// the 'Lazy Loading' phase.
  ///
  /// If true, the main axis extent of child will be cached as
  /// index/(key, extent) pairs when the child first laid out. If the [Key] of
  /// the item widget changes after caching, the corresponding cached extent
  /// will be removed and the new extent will be cached after layout.
  ///
  /// If false, the main axis extents of children will not be cached and all the
  /// previously cached data will be cleared.
  ///
  /// The default is false.
  ///
  /// This caching mechanism is suitable for scenarios where the main axis extents
  /// of children do not change or change infrequently during its life cycle.
  /// Using cached extent dimensions will improve scrolling performance,
  /// especially in scenarios that the scroll position changes drastically,
  /// as it will significantly reduce the layout times if the children
  /// have been laid out before.
  ///
  /// If the main axis extent of child at a certain index changes, the [Key] of
  /// the item widget needs to change to drive the framework to invalidate
  /// this child's cached value.
  ///
  /// All cached data will be cleared automatically if [scrollDirection] or
  /// [SliverConstraints.crossAxisExtent] changed.
  ///
  /// This has no effect if the extent of the children is predetermined,
  /// for example, the [itemExtent], [itemExtentBuilder] or [prototypeItem] have
  /// a non-null value, as they are already efficient.
  /// {@endtemplate}
  final bool enableItemExtentsCaching;
```

## Performance：(Window PC, Prefile mode, Scrolling from Top to Bottom with linear curves in one second)
### Without caching (max 13.4ms/frame, avg 10.1ms/frame)
![Image_20230918093837](https://github.com/flutter/flutter/assets/61075224/74c18d8e-cbc3-4b4d-adc9-fc3ebee6e5fb)

### Enable caching (max 6.5ms/frame, avg 4.9ms/frame)
![Image_20230918094106](https://github.com/flutter/flutter/assets/61075224/5450fba7-af47-49f9-994e-a6fe39fd406b)

## TODO:
1, Should we provide a method for users to proactively clear all the cached data(caching enabled)? I'm not so sure.
2, Add tests for this new feature.

## Test Demo Code
```dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
final scrollController = ScrollController(initialScrollOffset: childCount * 30);
final childCount = 10000;
void main() {
  runApp(
    MaterialApp(
      showPerformanceOverlay: true,
      scrollBehavior: MyScrollBehavior(),
      home: Scaffold(
        body: const ExampleApp(),
        floatingActionButton: FloatingActionButton(onPressed: () {
          if (scrollController.offset < 100) {
            scrollController.animateTo(
              childCount * 20,
                duration: const Duration(seconds: 1),
                curve: Curves.linear,
            );
          } else {
            scrollController.animateTo(
              0,
              duration: const Duration(seconds: 1),
              curve: Curves.linear,
            );
          }
        },),
      ),
    ),
  );
}

class ExampleApp extends StatefulWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return ExampleAppState();
  }
}

class ExampleAppState extends State<ExampleApp> {

  @override
  Widget build(BuildContext context) {
    return Scrollbar(
      thumbVisibility: true,
      controller: scrollController,
      child: ListView.builder(
        cacheExtent: 250,
        // scrollDirection: Axis.horizontal,
        controller: scrollController,
        itemCount: childCount,
        enableItemExtentsCaching: true,
        // itemExtent: 25.0,
        // itemExtentBuilder: (int index, SliverLayoutDimensions dimensions) {
        //   return 10;
        // },
        // itemExtentBuilder: (int index, SliverLayoutDimensions dimensions) {
        //   if (index % 2 == 0) {
        //     return 50;
        //   } else {
        //     return 100;
        //   }
        // },
        itemBuilder: (BuildContext context, int index) {
          var color = Colors.yellow;
          if (index % 2 == 0) {
            color = Colors.red;
          }

          return ColoredBox(
            color: color,
            child: Center(
              child: Text('Item $index'),
            ),
          );
        },
      ),
    );
  }
}

class MyScrollBehavior extends MaterialScrollBehavior {
  @override
  Widget buildScrollbar(
      BuildContext context, Widget child, ScrollableDetails details) {
    return child;
  }

  @override
  Set<PointerDeviceKind> get dragDevices => <PointerDeviceKind>{
    PointerDeviceKind.touch,
    PointerDeviceKind.mouse,
  };
}

```

xu-baolin committed:
[Performance] Introduce a caching mechanism to ListView to reduce unnecessary layout during scrolling

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

Piinks commented on packages/flutter/lib/src/rendering/sliver_list.dart:
```diff
@@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
+    _childrenWithoutLayout.clear();
+    _childConstraints = constraints.asBoxConstraints();
+    // Clear the cached extents if [crossAxisExtent] or [axis] changed.
+    if (constraints.crossAxisExtent != _lastCrossAxisExtent ||
```
viewportMainAxisExtent would probably clear the cache too.

xu-baolin replied:
Generally, the layout of items in the main axis direction is not constrained, so the cache is not cleared. At the same time, developers can explicitly clear the cache data through `invalidCachedItemExtents`.
Maybe both behaviors are okay and I can accept both.

Piinks replied:
> Generally, the layout of items in the main axis direction is not constrained

Totally! You are right. :)
Some folks size/position their children based on the viewport dimensions, especially not that we provide that information in the itemExtentBuilder feature you added recently. So it would probably be good to use a change in the viewport dimension as a cache invalidator.

Piinks reviewed (commented):
Mainly thoughts and questions here first, I am still wrapping my head around this and all the possible edge cases. I am very excited to see this though! :)

xu-baolin commented on packages/flutter/lib/src/rendering/sliver_list.dart:
```diff
@@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
+  /// previously cached data will be cleared.
+  ///
+  /// The cached data will be used to efficiently determine the target children
+  /// within the lazy loading scope, which helps improve layout performance.
```
Good point!
I considered exposing a property named like "Set<int> invalidCachedItemExtent" to let the render object delete the corresponding cached data.
But this seems to lack grace and cannot invalidate all cached values easily. Do you have any good suggestions?

Piinks replied:
Hmm, we could rely on the SliverChildDelegate.shouldRebuild to help invalidate the cache. In the SliverChildListDelegate that would work, but in the SliverChildBuilderDelegate it would always invalidate.. Maybe @chunhtai has some thoughts on this. Particularly with the child reordering edge cases, I think @chunhtai has some familiarity. :)

chunhtai replied:
I am not sure how this would work, even if the `SliverChildListDelegate` doesn't change, the child itself can still change size.

Piinks replied:
Oh right! That get's me every time. It is very common for folks to change size of scrollable children through animations or in response to the scroll position changing.

xu-baolin commented on packages/flutter/lib/src/rendering/sliver_list.dart:
```diff
@@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
+          if (super.paintExtentOf(child!) != _cachedItemExtents[index]!.mainAxisExtent) {
+            throw FlutterError(
+              'The item(index: $index) main axis extent has changed after caching, if this is '
+              'expected, change its Key to update the cached value.',
```
```Changing the key might not be an option in some cases.```
You are right, we should provide a way to invalidate certain caches.

```If the extent is different, can we just update it?```
Maybe we should throw here because we've used the wrong cached value to find the target position.

chunhtai replied:
This seems dangerous if developers are not sure what they are doing. It is also a bit confusing that it would require manual invalidation of caches. I think what would happen is developer will just use the invalidating caches everywhere so that it doesn't crash.

xu-baolin commented on packages/flutter/lib/src/rendering/sliver_list.dart:
```diff
@@ -236,17 +392,14 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
       if (!inLayoutRange) {
         if (child == null || indexOf(child!) != index) {
           // We are missing a child. Insert it (and lay it out) if possible.
```
This means that the missing child has been laid out before (removed from the tree due to offstage), and using the cached range will save layout costs.

Piinks reviewed (commented):
I've looped in @chunhtai for his thoughts on this. Thanks for your work here!

chunhtai reviewed (commented):
This api seems to be a bit dangerous. It is hard to predict child size change. I am not sure if this is something we want to pursue. There is also a PreferredSizeWidget, but that will require developer to give the size beforehand.

xu-baolin commented:
This is my thoughts.
This feature is enabled optional by the developer and only applies if the developer is sure that the children's sizes will not changed during the life cycle. I think there are many such scenarios, for example, on the web, once the child renders, the size will not change unless the user refreshes the page. In these scenarios, caching can greatly improve scrolling performance.






________________________________
来自TT移动邮箱 / From TT mailbox





________________________________
发件人: chunhtai ***@***.***>
发送时间: 2023年10月13日 03:53
收件人: flutter/flutter ***@***.***>
抄送: 徐宝林(Ava) ***@***.***>; Author ***@***.***>
主题: Re: [flutter/flutter] [New feature][Performance] Introduce a caching mechanism to ListView to reduce unnecessary layout during scrolling (PR #134864)


@chunhtai commented on this pull request.

This api seems to be a bit dangerous. It is hard to predict child size change. I am not sure if this is something we want to pursue. There is also a PreferredSizeWidget, but that will require developer to give the size beforehand.

________________________________

In packages/flutter/lib/src/rendering/sliver_list.dart<https://github.com/flutter/flutter/pull/134864#discussion_r1357281415>:

> @@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
   /// the main axis.
   RenderSliverList({
     required super.childManager,
-  });
+    bool enableItemExtentsCaching = false,
+  }) : _enableItemExtentsCaching = enableItemExtentsCaching;
+
+  /// Whether to cache the main axis extents of all laid out children.
+  ///
+  /// If true, the main axis extent of child will be cached as
+  /// index/extent pairs when the child first laid out.
+  ///
+  /// If false, the main axis extents of children will not be cached and
+  /// previously cached data will be cleared.
+  ///
+  /// The cached data will be used to efficiently determine the target children
+  /// within the lazy loading scope, which helps improve layout performance.


I am not sure how this would work, even if the SliverChildListDelegate doesn't change, the child itself can still change size.

________________________________

In packages/flutter/lib/src/rendering/sliver_list.dart<https://github.com/flutter/flutter/pull/134864#discussion_r1357292493>:

> +  }
+
+  void _performLoadingScopeChildrenLayoutIfNeeded() {
+    if (!enableItemExtentsCaching) {
+      return;
+    }
+    for (RenderBox? child = firstChild; child != null ; child = childAfter(child)) {
+      final int index = indexOf(child);
+      if (_childrenWithoutLayout.contains(index)) {
+        child.layout(_childConstraints, parentUsesSize: true);
+        // Throw if the real extent differs from the cached value.
+        assert(() {
+          if (super.paintExtentOf(child!) != _cachedItemExtents[index]!.mainAxisExtent) {
+            throw FlutterError(
+              'The item(index: $index) main axis extent has changed after caching, if this is '
+              'expected, change its Key to update the cached value.',


This seems dangerous if developers are not sure what they are doing. It is also a bit confusing that it would require manual invalidation of caches. I think what would happen is developer will just use the invalidating caches everywhere so that it doesn't crash.

―
Reply to this email directly, view it on GitHub<https://github.com/flutter/flutter/pull/134864#pullrequestreview-1674971562>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOR66GDACE653AFWYO4JCMLX7BDFJANCNFSM6AAAAAA42VIBIE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

________________________________
OPPO

本电子邮件及其附件含有OPPO公司的保密信息，仅限于邮件指明的收件人（包含个人及群组）使用。禁止任何人在未经授权的情况下以任何形式使用。如果您错收了本邮件，切勿传播、分发、复制、印刷或使用本邮件之任何部分或其所载之任何内容，并请立即以电子邮件通知发件人并删除本邮件及其附件。
网络通讯固有缺陷可能导致邮件被截留、修改、丢失、破坏或包含计算机病毒等不安全情况，OPPO对此类错误或遗漏而引致之任何损失概不承担责任并保留与本邮件相关之一切权利。
除非明确说明，本邮件及其附件无意作为在任何国家或地区之要约、招揽或承诺，亦无意作为任何交易或合同之正式确认。 发件人、其所属机构或所属机构之关联机构或任何上述机构之股东、董事、高级管理人员、员工或其他任何人（以下称“发件人”或“OPPO”）不因本邮件之误送而放弃其所享之任何权利，亦不对因故意或过失使用该等信息而引发或可能引发的损失承担任何责任。
文化差异披露：因全球文化差异影响，单纯以YES\OK或其他简单词汇的回复并不构成发件人对任何交易或合同之正式确认或接受，请与发件人再次确认以获得明确书面意见。发件人不对任何受文化差异影响而导致故意或错误使用该等信息所造成的任何直接或间接损害承担责任。
This e-mail and its attachments contain confidential information from OPPO, which is intended only for the person or entity whose address is listed above. Any use of the information contained herein in any way (including, but not limited to, total or partial disclosure, reproduction, or dissemination) by persons other than the intended recipient(s) is prohibited. If you are not the intended recipient, please do not read, copy, distribute, or use this information. If you have received this transmission in error, please notify the sender immediately by reply e-mail and then delete this message.
Electronic communications may contain computer viruses or other defects inherently, may not be accurately and/or timely transmitted to other systems, or may be intercepted, modified ,delayed, deleted or interfered. OPPO shall not be liable for any damages that arise or may arise from such matter and reserves all rights in connection with the email.
Unless expressly stated, this e-mail and its attachments are provided without any warranty, acceptance or promise of any kind in any country or region, nor constitute a formal confirmation or acceptance of any transaction or contract. The sender, together with its affiliates or any shareholder, director, officer, employee or any other person of any such institution (hereinafter referred to as "sender" or "OPPO") does not waive any rights and shall not be liable for any damages that arise or may arise from the intentional or negligent use of such information.
Cultural Differences Disclosure: Due to global cultural differences, any reply with only YES\OK or other simple words does not constitute any confirmation or acceptance of any transaction or contract, please confirm with the sender again to ensure clear opinion in written form. The sender shall not be responsible for any direct or indirect damages resulting from the intentional or misuse of such information.

chunhtai commented:
I can see how it can be useful in some circumstances, but I am still not convince we should provide it in the framework as it can be easily misused. Have you considered put this into a third party package instead?

xu-baolin commented:
> I can see how it can be useful in some circumstances, but I am still not convince we should provide it in the framework as it can be easily misused. Have you considered put this into a third party package instead?

- `3rd party package`
If this has no benefits or brings side effects to the framework, or there are only very rare benefit scenarios, third-party package is an option. Or we can continue to discuss this topic in depth and collect developer voices.
Even as a 3rd package, we also need resolve the misusage before published, right? : )

- `Misusage`
In addition to complete documentation, some necessary checks in debug mode can well catch the misusage cases.

- `manual invalidation of caches`
Currently, I think we should only support a property that invalidate all cached data. When explicit size changes occur (such as reordering, adding or removing a child from the middle), even if the user forgets to manually clear it, the framework will throw an assertion.





________________________________
来自TT移动邮箱 / From TT mailbox





________________________________
发件人: chunhtai ***@***.***>
发送时间: 2023年10月13日 06:52
收件人: flutter/flutter ***@***.***>
抄送: 徐宝林(Ava) ***@***.***>; Author ***@***.***>
主题: Re: [flutter/flutter] [New feature][Performance] Introduce a caching mechanism to ListView to reduce unnecessary layout during scrolling (PR #134864)


I can see how it can be useful in some circumstances, but I am still not convince we should provide it in the framework as it can be easily misused. Have you considered put this into a third party package instead?

―
Reply to this email directly, view it on GitHub<https://github.com/flutter/flutter/pull/134864#issuecomment-1760476148>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOR66GCRCXAWZIHCM5YMXCDX7BYB7ANCNFSM6AAAAAA42VIBIE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

________________________________
OPPO

本电子邮件及其附件含有OPPO公司的保密信息，仅限于邮件指明的收件人（包含个人及群组）使用。禁止任何人在未经授权的情况下以任何形式使用。如果您错收了本邮件，切勿传播、分发、复制、印刷或使用本邮件之任何部分或其所载之任何内容，并请立即以电子邮件通知发件人并删除本邮件及其附件。
网络通讯固有缺陷可能导致邮件被截留、修改、丢失、破坏或包含计算机病毒等不安全情况，OPPO对此类错误或遗漏而引致之任何损失概不承担责任并保留与本邮件相关之一切权利。
除非明确说明，本邮件及其附件无意作为在任何国家或地区之要约、招揽或承诺，亦无意作为任何交易或合同之正式确认。 发件人、其所属机构或所属机构之关联机构或任何上述机构之股东、董事、高级管理人员、员工或其他任何人（以下称“发件人”或“OPPO”）不因本邮件之误送而放弃其所享之任何权利，亦不对因故意或过失使用该等信息而引发或可能引发的损失承担任何责任。
文化差异披露：因全球文化差异影响，单纯以YES\OK或其他简单词汇的回复并不构成发件人对任何交易或合同之正式确认或接受，请与发件人再次确认以获得明确书面意见。发件人不对任何受文化差异影响而导致故意或错误使用该等信息所造成的任何直接或间接损害承担责任。
This e-mail and its attachments contain confidential information from OPPO, which is intended only for the person or entity whose address is listed above. Any use of the information contained herein in any way (including, but not limited to, total or partial disclosure, reproduction, or dissemination) by persons other than the intended recipient(s) is prohibited. If you are not the intended recipient, please do not read, copy, distribute, or use this information. If you have received this transmission in error, please notify the sender immediately by reply e-mail and then delete this message.
Electronic communications may contain computer viruses or other defects inherently, may not be accurately and/or timely transmitted to other systems, or may be intercepted, modified ,delayed, deleted or interfered. OPPO shall not be liable for any damages that arise or may arise from such matter and reserves all rights in connection with the email.
Unless expressly stated, this e-mail and its attachments are provided without any warranty, acceptance or promise of any kind in any country or region, nor constitute a formal confirmation or acceptance of any transaction or contract. The sender, together with its affiliates or any shareholder, director, officer, employee or any other person of any such institution (hereinafter referred to as "sender" or "OPPO") does not waive any rights and shall not be liable for any damages that arise or may arise from the intentional or negligent use of such information.
Cultural Differences Disclosure: Due to global cultural differences, any reply with only YES\OK or other simple words does not constitute any confirmation or acceptance of any transaction or contract, please confirm with the sender again to ensure clear opinion in written form. The sender shall not be responsible for any direct or indirect damages resulting from the intentional or misuse of such information.

chunhtai commented:
> Even as a 3rd package, we also need resolve the misusage before published, right? : )

I propose as 3rd party because it is has a limited scope.

> In addition to complete documentation, some necessary checks in debug mode can well catch the misusage cases.

My concern are:
1. It is hard to predict child would change size or not. Developer may think their widget has fixed size while it may not, and they use this widget without knowing it
2. We only know things are wrong (thus throw or assert) when the size did change. That means if a widget only change size in a corner case, it will be hard to catch this kind of bug during developerment.

xu-baolin commented:
That makes sense, thanks for your reviews.

________________________________
来自TT移动邮箱 / From TT mailbox


________________________________
发件人: chunhtai ***@***.***>
发送时间: 2023年10月13日 06:52
收件人: flutter/flutter ***@***.***>
抄送: 徐宝林(Ava) ***@***.***>; Author ***@***.***>
主题: Re: [flutter/flutter] [New feature][Performance] Introduce a caching mechanism to ListView to reduce unnecessary layout during scrolling (PR #134864)


I can see how it can be useful in some circumstances, but I am still not convince we should provide it in the framework as it can be easily misused. Have you considered put this into a third party package instead?

―
Reply to this email directly, view it on GitHub<https://github.com/flutter/flutter/pull/134864#issuecomment-1760476148>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOR66GCRCXAWZIHCM5YMXCDX7BYB7ANCNFSM6AAAAAA42VIBIE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

________________________________
OPPO

本电子邮件及其附件含有OPPO公司的保密信息，仅限于邮件指明的收件人（包含个人及群组）使用。禁止任何人在未经授权的情况下以任何形式使用。如果您错收了本邮件，切勿传播、分发、复制、印刷或使用本邮件之任何部分或其所载之任何内容，并请立即以电子邮件通知发件人并删除本邮件及其附件。
网络通讯固有缺陷可能导致邮件被截留、修改、丢失、破坏或包含计算机病毒等不安全情况，OPPO对此类错误或遗漏而引致之任何损失概不承担责任并保留与本邮件相关之一切权利。
除非明确说明，本邮件及其附件无意作为在任何国家或地区之要约、招揽或承诺，亦无意作为任何交易或合同之正式确认。 发件人、其所属机构或所属机构之关联机构或任何上述机构之股东、董事、高级管理人员、员工或其他任何人（以下称“发件人”或“OPPO”）不因本邮件之误送而放弃其所享之任何权利，亦不对因故意或过失使用该等信息而引发或可能引发的损失承担任何责任。
文化差异披露：因全球文化差异影响，单纯以YES\OK或其他简单词汇的回复并不构成发件人对任何交易或合同之正式确认或接受，请与发件人再次确认以获得明确书面意见。发件人不对任何受文化差异影响而导致故意或错误使用该等信息所造成的任何直接或间接损害承担责任。
This e-mail and its attachments contain confidential information from OPPO, which is intended only for the person or entity whose address is listed above. Any use of the information contained herein in any way (including, but not limited to, total or partial disclosure, reproduction, or dissemination) by persons other than the intended recipient(s) is prohibited. If you are not the intended recipient, please do not read, copy, distribute, or use this information. If you have received this transmission in error, please notify the sender immediately by reply e-mail and then delete this message.
Electronic communications may contain computer viruses or other defects inherently, may not be accurately and/or timely transmitted to other systems, or may be intercepted, modified ,delayed, deleted or interfered. OPPO shall not be liable for any damages that arise or may arise from such matter and reserves all rights in connection with the email.
Unless expressly stated, this e-mail and its attachments are provided without any warranty, acceptance or promise of any kind in any country or region, nor constitute a formal confirmation or acceptance of any transaction or contract. The sender, together with its affiliates or any shareholder, director, officer, employee or any other person of any such institution (hereinafter referred to as "sender" or "OPPO") does not waive any rights and shall not be liable for any damages that arise or may arise from the intentional or negligent use of such information.
Cultural Differences Disclosure: Due to global cultural differences, any reply with only YES\OK or other simple words does not constitute any confirmation or acceptance of any transaction or contract, please confirm with the sender again to ensure clear opinion in written form. The sender shall not be responsible for any direct or indirect damages resulting from the intentional or misuse of such information.

chunhtai commented:
I am going to close this PR. Still thanks for your contribution even though we decide to not move forward.

yakagami commented:
I understand the PR is already closed, but could it make sense to add this feature only where the extent is already known? Eg. with a fixed extent list, prototype extent list or the new `SliverVariedExtentList`? These `Sliver`s are meant for performance and would already cover a lot of cases as most `List`s inherently have children with identical extents.

knopp commented:
@xu-baolin, just a heads-up, we do have a package that does similar thing and it performs well. https://superlistapp.github.io/super_sliver_list/

Cache invalidation is generally not a problem, since cached extents are always treated as estimations, every time items in the cache area are laid out cached extents are estimated and scroll offset / extends are corrected to reflect the difference.

Just mentioning this to possibly save you some time.

fyxtc commented:
@knopp Hi, does your package solve this issue? https://github.com/flutter/flutter/issues/143817#issuecomment-1969523197

knopp commented:
> @knopp Hi, does your package solve this issue? [#143817 (comment)](https://github.com/flutter/flutter/issues/143817#issuecomment-1969523197)

Unlikely. The problem there seems to be that the items are laid out with small extent initially and then expanded as the image loads (asynchronously).

fyxtc commented:
> > @knopp Hi, does your package solve this issue? [#143817 (comment)](https://github.com/flutter/flutter/issues/143817#issuecomment-1969523197)
> 
> Unlikely. The problem there seems to be that the items are laid out with small extent initially and then expanded as the image loads (asynchronously).

Yes, it's a very serious problem for my app, the latest version of flutter still hasn't solved it, the related issue is here https://github.com/flutter/flutter/issues/99158, if you can solve it, it would be extremely helpful and I'm more than happy to donate!

knopp commented:
> > > @knopp Hi, does your package solve this issue? [#143817 (comment)](https://github.com/flutter/flutter/issues/143817#issuecomment-1969523197)
> > 
> > 
> > Unlikely. The problem there seems to be that the items are laid out with small extent initially and then expanded as the image loads (asynchronously).
> 
> Yes, it's a very serious problem for my app, the latest version of flutter still hasn't solved it, the related issue is here #99158, if you can solve it, it would be extremely helpful and I'm more than happy to donate!

You can always fix this by caching the size of images yourself and then use SizedBox with cached size when buildings item while scrolling back. The main issue here is that your initially create all your items with very small extent and then they resize after a while when the image is loaded. No ListView implementation can fix this (in theory resizing items in cache area can correct the scroll offset transparently but this will fall apart when you scroll too quickly so the best way is to always create the widget with proper size once you know it).

fyxtc commented:
@knopp Very helpful, thank you
